### PR TITLE
Dry types

### PR DESCRIPTION
**Pull request by sagacic-tim**
Everything is ready to go. Want to move on to dry_implementation

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -8,7 +8,8 @@ ruby "3.2.2"
 # onment is bootstrapped.”
 gem 'dotenv-rails', :groups => [:development, :test]
 
-# Bundle edge Rails instead: gem "rails", github: "rails/rails", branch: "main"
+# Bundle edge Rails instead: gem "rails", github: "rails/rails",
+# branch: "main"
```
Comment (rAlexander89) [inline Gemfile]: remove

File: app/controllers/api/v1/friends_controller.rb [hunk 2]
```diff
@@ -1,20 +1,28 @@
-class Api::V1::FriendsController < ApplicationController 
+class Api::V1::FriendsController < ApplicationController
+  require 'debug'
```
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb:2]: remove. this is already globally available.

File: app/controllers/api/v1/friends_controller.rb [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
     end
   end
+  # create a new friend Post request
   def create
     @friend = Friend.new(friend_params)
-
+    # debugger
+    # binding.break 
+    # logger.debug "New post: #{@post.attributes.inspect}"
+    # .debug "Post should be valid: #{@post.valid?}"
```
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb]: remove. Cruft.

Fun fact, "Cruft" is an ironic play on the word "craft"


```
Cruft is a jargon word for anything that is left over, redundant and getting in the way. It is used particularly for defective, superseded, useless, superfluous, or dysfunctional elements in computer software. 
```

File: app/controllers/api/v1/friends_controller.rb [hunk 4]
```diff
@@ -1,20 +1,28 @@
...
     if @friend.save
-      render json: {status: "SUCCESS", message: "Amazing, a new friend was created!", data: friend}, status: :created
+      # binding.break
```
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb]: cruft

File: app/controllers/api/v1/friends_controller.rb [hunk 5]
```diff
@@ -1,20 +1,28 @@
...
-
+    # debugger
+    # binding.break 
+    # logger.debug "New post: #{@post.attributes.inspect}"
+    # .debug "Post should be valid: #{@post.valid?}"
     if @friend.save
-      render json: {status: "SUCCESS", message: "Amazing, a new friend was created!", data: friend}, status: :created
+      # binding.break 
+      render json: {status: "SUCCESS", message: "Amazing, a new friend was created!", data: @friend}, status: :created
+      # logger.debug "Post was successful: #{@post.valid?}"
```
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb]: cruft

File: app/controllers/api/v1/friends_controller.rb [hunk 6]
```diff
@@ -1,20 +1,28 @@
...
     else
-      render json: friend.errors, status: :unprocessable_entity
+      # logger.debug "Post has ewrror: #{@post.valid?}"
+      # binding.break
```
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb]: cruft

File: app/models/friend.rb [hunk 7]
```diff
@@ -5,29 +5,26 @@
 require "smartystreets_ruby_sdk/us_street/match_type"
 # require 'dry/types'
 # require_relative '../lib/tasks/decimal_types'
+# require 'colorize' # Extends String class
+# require 'colorized_string' # add ColorizedString class
```
Comment (rAlexander89) [inline app/models/friend.rb:9]: cruft.

I get that you want to use these, but they should be introduced in the branch that introduces that functionality.


PRs (pull requests) should be constrained to single issues to keep code changes clean and understandable.

File: app/models/friend.rb [hunk 8]
```diff
@@ -5,29 +5,26 @@
...
     serialize :verification_info
 
-    private
+    before_validation :validate_address_with_smartystreets
 
-    def show_full_name
-        full_name ="#{name_title} #{name_first} #{name_middle} #{name_last} #{name_suffix}"
-        puts "\nFull Name: " + full_name + "\n\n"
-    end
+    # private
```
Comment (rAlexander89) [inline app/models/friend.rb]: cruft

File: app/models/friend.rb [hunk 9]
```diff
@@ -58,7 +55,18 @@ def validate_address_with_smartystreets
...
+        elsif lookup.match == :exact
+            puts "There is at least one candidate.\n\nThe match parameter IS set to " + lookup.match.to_s + ", this indicates an exact match.\n\n" 
+        elsif lookup.match == :range
+            puts "There is at least one candidate.\n\nThe match parameter IS set to " + lookup.match.to_s + ", this Indicates a match within a range of addresses.\n\n" 
+        elsif lookup.match == :invalid
+            puts "There is at least one candidate.\n\nThe match parameter IS set to " + lookup.match.to_s + ", this means the addrress may or may not be valid and you need to perform further validations.\n\n"
+        else
+            puts "Invalid match type" + "\n\n"
+        end
+
```
Comment (rAlexander89) [inline app/models/friend.rb]: Cruft. This is not needed. If you need to know wether an address is as you expect it, return an error or a success.

File: app/controllers/api/v1/friends_controller.rb (hunk 2 above)
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb:2]: remove. this is already globally available.

File: app/controllers/api/v1/friends_controller.rb [hunk 10]
```diff
@@ -1,20 +1,28 @@
...
   def index
     friends = Friend.all 
-
     if friends
       render json: {status: "SUCCESS", message: "Fetched all the friends you got", data: friends}, status: :ok
     else
       render json: friends.errors, status: :bad_request
     end
   end
+  # create a new friend Post request
```
Comment (rAlexander89) [inline app/controllers/api/v1/friends_controller.rb]: cruft

File: Gemfile [hunk 11]
```diff
@@ -38,8 +39,9 @@ gem "bootsnap", require: false
 # Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
 # gem "image_processing", "~> 1.2"
 
-# Use Rack CORS for handling Cross-Origin Resource Sharing (CORS), making cross-origin AJAX possible
-# gem "rack-cors"
+# Rack::Cors provides support for Cross-Origin Resource
+# Sharing (CORS) for Rack compatible web applications.
+gem 'rack-cors', '~> 1.1', '>= 1.1.1'
```
Comment (sagacic-tim) [inline Gemfile:43]: This is the way to include rack-cors. Deletd the duplicate

File: Gemfile [hunk 12]
```diff
@@ -103,13 +101,23 @@ gem 'dry-types', '~> 1.7'
...
 group :development, :test do
-  # See https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
-  gem "debug", platforms: %i[ mri mingw x64_mingw ]
+  # This library provides debugging functionality to Ruby (MRI) 2.6
+  # and later.
+  # This debug.rb is replacement of traditional lib/debug.rb
+  # standard library which is implemented by set_trace_func. New
+  #debug.rb has several advantages:
+  # See: https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
+  gem "debug", ">= 1.0.0"
```
Comment (sagacic-tim) [inline Gemfile:115]: Enhanced debugging

File: Gemfile [hunk 13]
```diff
@@ -103,13 +101,23 @@ gem 'dry-types', '~> 1.7'
...
 end
 
 group :development do
   # Speed up commands on slow machines / big apps [https://github.com/rails/spring]
   # gem "spring"
-  gem 'annotate'
+  gem 'annotate', git: 'https://github.com/ctran/annotate_models.git'
```
Comment (sagacic-tim) [inline Gemfile:121]: The line gem 'annotate', git: 'https://github.com/ctran/annotate_models.git' in your Gemfile indicates that you want to include the annotate_models gem from the specified GitHub repository

File: app/models/friend.rb (hunk 7 above)
Comment (sagacic-tim) [reply app/models/friend.rb:9]: I am not going to back track at this point. I will store that bit of knowledge away for future api's like the rewrite of this one.

**Comment by sagacic-tim** [comment]
Everything ready to go. I want to move on to dry_implementation which should represent the final piece of the back end work.

**Review by rAlexander89 — APPROVED** [review]
lgtm

**Pull request merged**